lib/src/library_scanner.dart

```diff
@@ -40,40 +38,34 @@ class LibraryScanner {
   LibraryScanner._(
       this.packageName, this._packagePath, this._session, this._overrides);
 
-  factory LibraryScanner(String dartSdkPath, String packagePath,
-      {List<LibraryOverride> overrides}) {
+  static Future<LibraryScanner> create(String dartSdkPath, String packagePath,
+      {List<LibraryOverride> overrides}) async {
     var dotPackagesPath = p.join(packagePath, '.packages');
     if (!FileSystemEntity.isFileSync(dotPackagesPath)) {
       throw StateError('A package configuration file was not found at the '
           'expected location.\n$dotPackagesPath');
     }
-    var dotPackagesBytes = File(dotPackagesPath).readAsBytesSync();
-    var packageMap =
-        package_config.parse(dotPackagesBytes, p.toUri(dotPackagesPath));
+    final dotPackagesFile = File(dotPackagesPath);
+    final config = await package_config.loadPackageConfig(dotPackagesFile);
 
     String package;
-    var packageNames = <String>[];
-    packageMap.forEach((k, v) {
-      if (package != null) {
-        return;
-      }
-
-      assert(package == null);
-
-      // if there is an exact match to the lib directory, use that
-      if (p.fromUri(v) == p.join(packagePath, 'lib/')) {
-        package = k;
-      }
-
-      if (p.isWithin(packagePath, p.fromUri(v))) {
-        packageNames.add(k);
+    final packageNames = <String>[];
+    config.packages.forEach((pkg) {
+      if (package != null) return;
+
+      final pkgRootFilePath = Directory.fromUri(pkg.root).path;
+      if (pkgRootFilePath == '$packagePath${Platform.pathSeparator}') {
```

Comment:
Wouldn't it be safer to normalize paths and compare them?

Reply:
That's the normalization right there. :)
Do you know any method in `package:path` that takes a directory path and returns it without the end file separator? e.g. `a/b/c/` -> `a/b/c`